From 003deecd1140c5e8d8b06ee44985aeda8015cb64 Mon Sep 17 00:00:00 2001
From: Emmanuel Blot <emmanuel.blot@free.fr>
Date: Fri, 7 Aug 2026 11:10:41 +0200
Subject: [PATCH] hw/arm: bletchley: align the I2C layout with the device tree

Several I2C devices described by the board device tree were absent or
mismodelled: the per-sled FRU EEPROMs and PCA9536 expanders on buses
0-5 were never instantiated, and the FRU EEPROMs on buses 6 and 7 were
sized as 24c512 parts instead of the 24c64 the device tree specifies.

Instantiate the missing devices at the correct sizes and, now that
models exist for them, wire the INA230 power monitors and the PCA9539
expanders directly instead of leaving the former unmodelled and
standing the latter in with a PCA9552.

Document the devices that still have no QEMU model in the file header,
drop a stale comment for a sensor absent from the device tree, and
remove the unused qapi/error.h include.

Signed-off-by: Emmanuel Blot <emmanuel.blot@free.fr>
---
 hw/arm/aspeed_ast2600_bletchley.c | 27 ++++++++++++++++++++-------
 1 file changed, 20 insertions(+), 7 deletions(-)

diff --git a/hw/arm/aspeed_ast2600_bletchley.c b/hw/arm/aspeed_ast2600_bletchley.c
index 370baaa037..7d0b4afefb 100644
--- a/hw/arm/aspeed_ast2600_bletchley.c
+++ b/hw/arm/aspeed_ast2600_bletchley.c
@@ -4,17 +4,27 @@
  * Copyright 2016 IBM Corp.
  *
  * SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ * Current limitations: the following devices from the device tree have no
+ * QEMU model and are therefore not emulated:
+ *   - mps,mp5023 power monitors
+ *   - fcs,fusb302 USB Type-C PD controllers
+ *   - adi,adm1278 hot-swap controller
+ *   - infineon,slb9670 TPM (SPI)
+ *   - ipmb-dev IPMB interface
  */
 
 #include "qemu/osdep.h"
-#include "qapi/error.h"
 #include "hw/arm/machines-qom.h"
 #include "hw/arm/aspeed.h"
 #include "hw/arm/aspeed_soc.h"
 #include "hw/gpio/pca9552.h"
+#include "hw/gpio/pca9554.h"
 #include "hw/nvram/eeprom_at24c.h"
 #include "hw/rtc/ds1338.h"
+#include "hw/sensor/ina230.h"
 #include "hw/sensor/tmp421.h"
+
 /* Bletchley hardware value */
 #define BLETCHLEY_BMC_HW_STRAP1 0x00002000
 #define BLETCHLEY_BMC_HW_STRAP2 0x00000801
@@ -33,30 +43,30 @@ static void bletchley_bmc_i2c_init(AspeedMachineState *bmc)
 
     /* Bus 0 - 5 all have the same config. */
     for (int i = 0; i < 6; i++) {
-        /* Missing model: ti,ina230 @ 0x45 */
+        i2c_slave_create_simple(i2c[i], TYPE_INA230, 0x45);
         /* Missing model: mps,mp5023 @ 0x40 */
         i2c_slave_create_simple(i2c[i], TYPE_TMP421, 0x4f);
-        /* Missing model: nxp,pca9539 @ 0x76, but PCA9552 works enough */
-        i2c_slave_create_simple(i2c[i], TYPE_PCA9552, 0x76);
+        i2c_slave_create_simple(i2c[i], TYPE_PCA9536, 0x41);
+        i2c_slave_create_simple(i2c[i], TYPE_PCA9539, 0x76);
         i2c_slave_create_simple(i2c[i], TYPE_PCA9552, 0x67);
         /* Missing model: fsc,fusb302 @ 0x22 */
+        at24c_eeprom_init(i2c[i], 0x54, 8 * KiB);
     }
 
     /* Bus 6 */
-    at24c_eeprom_init(i2c[6], 0x56, 65536);
+    at24c_eeprom_init(i2c[6], 0x56, 8 * KiB);
     /* Missing model: nxp,pcf85263 @ 0x51 , but ds1338 works enough */
     i2c_slave_create_simple(i2c[6], TYPE_DS1338, 0x51);
 
 
     /* Bus 7 */
-    at24c_eeprom_init(i2c[7], 0x54, 65536);
+    at24c_eeprom_init(i2c[7], 0x54, 8 * KiB);
 
     /* Bus 9 */
     i2c_slave_create_simple(i2c[9], TYPE_TMP421, 0x4f);
 
     /* Bus 10 */
     i2c_slave_create_simple(i2c[10], TYPE_TMP421, 0x4f);
-    /* Missing model: ti,hdc1080 @ 0x40 */
     i2c_slave_create_simple(i2c[10], TYPE_PCA9552, 0x67);
 
     /* Bus 12 */
@@ -64,6 +74,9 @@ static void bletchley_bmc_i2c_init(AspeedMachineState *bmc)
     i2c_slave_create_simple(i2c[12], TYPE_TMP421, 0x4c);
     i2c_slave_create_simple(i2c[12], TYPE_TMP421, 0x4d);
     i2c_slave_create_simple(i2c[12], TYPE_PCA9552, 0x67);
+
+    /* Bus 13 */
+    /* Missing model: ipmb-dev @ 0x10 */
 }
 
 static void aspeed_machine_bletchley_class_init(ObjectClass *oc,
-- 
2.53.0-Meta

